### PR TITLE
Support private networking for managed nodegroups

### PR DESCRIPTION
**Pull request by sayboras**
### Description

Fixes https://github.com/weaveworks/eksctl/issues/1748
Fixes https://github.com/weaveworks/eksctl/issues/1575

### Approach
- Add --node-private-networking cli flag for create cluster command
- Add privateNetwork configuration param for ManagedNodeGroup in config file

I take this chance to address lack of unit test cases for create cluster CLI command as well. 

### Testing
<details>
<summary>Create managed cluster with private networking cli</summary>

```
$ ./eksctl create cluster --managed --node-private-networking
[ℹ]  eksctl version 0.18.0-dev+b8b1a818.2020-04-17T21:40:08Z
[ℹ]  using region us-east-2
[ℹ]  setting availability zones to [us-east-2a us-east-2b us-east-2c]
[ℹ]  subnets for us-east-2a - public:192.168.0.0/19 private:192.168.96.0/19
[ℹ]  subnets for us-east-2b - public:192.168.32.0/19 private:192.168.128.0/19
[ℹ]  subnets for us-east-2c - public:192.168.64.0/19 private:192.168.160.0/19
[ℹ]  using Kubernetes version 1.15
[ℹ]  creating EKS cluster "attractive-creature-1587123638" in "us-east-2" region with managed nodes
[ℹ]  will create 2 separate CloudFormation stacks for cluster itself and the initial managed nodegroup
[ℹ]  if you encounter any issues, check CloudFormation console or try 'eksctl utils describe-stacks --region=us-east-2 --cluster=attractive-creature-1587123638'
[ℹ]  CloudWatch logging will not be enabled for cluster "attractive-creature-1587123638" in "us-east-2"
[ℹ]  you can enable it with 'eksctl utils update-cluster-logging --region=us-east-2 --cluster=attractive-creature-1587123638'
[ℹ]  Kubernetes API endpoint access will use default of {publicAccess=true, privateAccess=false} for cluster "attractive-creature-1587123638" in "us-east-2"
[ℹ]  2 sequential tasks: { create cluster control plane "attractive-creature-1587123638", create managed nodegroup "ng-b8cf7a66" }
[ℹ]  building cluster stack "eksctl-attractive-creature-1587123638-cluster"
[ℹ]  deploying stack "eksctl-attractive-creature-1587123638-cluster"
[ℹ]  building managed nodegroup stack "eksctl-attractive-creature-1587123638-nodegroup-ng-b8cf7a66"
[ℹ]  deploying stack "eksctl-attractive-creature-1587123638-nodegroup-ng-b8cf7a66"
[✔]  all EKS cluster resources for "attractive-creature-1587123638" have been created
[✔]  saved kubeconfig as "/home/tammach/.kube/config"
[ℹ]  nodegroup "ng-b8cf7a66" has 2 node(s)
[ℹ]  node "ip-192-168-174-114.us-east-2.compute.internal" is ready
[ℹ]  node "ip-192-168-98-180.us-east-2.compute.internal" is ready
[ℹ]  waiting for at least 2 node(s) to become ready in "ng-b8cf7a66"
[ℹ]  nodegroup "ng-b8cf7a66" has 2 node(s)
[ℹ]  node "ip-192-168-174-114.us-east-2.compute.internal" is ready
[ℹ]  node "ip-192-168-98-180.us-east-2.compute.internal" is ready
[ℹ]  kubectl command should work with "/home/tammach/.kube/config", try 'kubectl get nodes'
[✔]  EKS cluster "attractive-creature-1587123638" in "us-east-2" region is ready


$ aws ec2 describe-instances --instance-ids i-0802bfd85bc1fb72c | grep SubnetId
                    "SubnetId": "subnet-00a788781be10df8a",
                            "SubnetId": "subnet-00a788781be10df8a",

$ aws ec2 describe-subnets --subnet-ids subnet-00a788781be10df8a                        
{
    "Subnets": [
        {
            "AvailabilityZone": "us-east-2a",
            "AvailabilityZoneId": "use2-az1",
            "AvailableIpAddressCount": 8176,
            "CidrBlock": "192.168.96.0/19",
            "DefaultForAz": false,
            "MapPublicIpOnLaunch": false,
            "State": "available",
            "SubnetId": "subnet-00a788781be10df8a",
            "VpcId": "vpc-060aa9359ca87b956",
            "OwnerId": "363193625107",
            "AssignIpv6AddressOnCreation": false,
            "Ipv6CidrBlockAssociationSet": [],
            "Tags": [
                {
                    "Key": "aws:cloudformation:stack-name",
                    "Value": "eksctl-attractive-creature-1587123638-cluster"
                },
                {
                    "Key": "alpha.eksctl.io/eksctl-version",
                    "Value": "0.18.0-dev+b8b1a818.2020-04-17T21:40:08Z"
                },
                {
                    "Key": "kubernetes.io/role/internal-elb",
                    "Value": "1"
                },
                {
                    "Key": "aws:cloudformation:logical-id",
                    "Value": "SubnetPrivateUSEAST2A"
                },
                {
                    "Key": "eksctl.cluster.k8s.io/v1alpha1/cluster-name",
                    "Value": "attractive-creature-1587123638"
                },
                {
                    "Key": "Name",
                    "Value": "eksctl-attractive-creature-1587123638-cluster/SubnetPrivateUSEAST2A"
                },
                {
                    "Key": "alpha.eksctl.io/cluster-name",
                    "Value": "attractive-creature-1587123638"
                },
                {
                    "Key": "aws:cloudformation:stack-id",
                    "Value": "arn:aws:cloudformation:us-east-2:363193625107:stack/eksctl-attractive-creature-1587123638-cluster/44ac80e0-80a0-11ea-8891-0a3e4885fce2"
                },
                {
                    "Key": "kubernetes.io/cluster/attractive-creature-1587123638",
                    "Value": "shared"
                }
            ],
            "SubnetArn": "arn:aws:ec2:us-east-2:363193625107:subnet/subnet-00a788781be10df8a"
        }
    ]
}

```

</details>

<details>
<summary>Create a new nodegroup with private networking flag</summary>

```shell script
$ ./eksctl create nodegroup --managed --nodes=1 --nodes-min=1 --nodes-max=2 --cluster attractive-creature-1587123638 --node-private-networking
[ℹ]  eksctl version 0.18.0-dev+b8b1a818.2020-04-17T21:40:08Z
[ℹ]  using region us-east-2
[ℹ]  will use version 1.15 for new nodegroup(s) based on control plane version
[ℹ]  1 nodegroup(s) that already exist (ng-b8cf7a66) will be excluded
[ℹ]  1 nodegroup (ng-beadf89e) was included (based on the include/exclude rules)
[ℹ]  combined exclude rules: ng-b8cf7a66
[ℹ]  no nodegroups present in the current set were excluded by the filter
[ℹ]  will create a CloudFormation stack for each of 1 managed nodegroups in cluster "attractive-creature-1587123638"
[ℹ]  2 sequential tasks: { fix cluster compatibility, 1 task: { 1 task: { create managed nodegroup "ng-beadf89e" } } }
[ℹ]  checking cluster stack for missing resources
[ℹ]  cluster stack is missing resources for Fargate
[ℹ]  adding missing resources to cluster stack
[ℹ]  re-building cluster stack "eksctl-attractive-creature-1587123638-cluster"
[✔]  all resources in cluster stack "eksctl-attractive-creature-1587123638-cluster" are up-to-date
[ℹ]  building managed nodegroup stack "eksctl-attractive-creature-1587123638-nodegroup-ng-beadf89e"
[ℹ]  deploying stack "eksctl-attractive-creature-1587123638-nodegroup-ng-beadf89e"
[✔]  created 0 nodegroup(s) in cluster "attractive-creature-1587123638"
[ℹ]  nodegroup "ng-beadf89e" has 1 node(s)
[ℹ]  node "ip-192-168-142-112.us-east-2.compute.internal" is ready
[ℹ]  waiting for at least 1 node(s) to become ready in "ng-beadf89e"
[ℹ]  nodegroup "ng-beadf89e" has 1 node(s)
[ℹ]  node "ip-192-168-142-112.us-east-2.compute.internal" is ready
[✔]  created 1 managed nodegroup(s) in cluster "attractive-creature-1587123638"
[ℹ]  checking security group configuration for all nodegroups
[ℹ]  all nodegroups have up-to-date configuration

$ kgno
NAME                                            STATUS   ROLES    AGE     VERSION
ip-192-168-142-112.us-east-2.compute.internal   Ready    <none>   82s     v1.15.10-eks-bac369
ip-192-168-174-114.us-east-2.compute.internal   Ready    <none>   8m22s   v1.15.10-eks-bac369
ip-192-168-98-180.us-east-2.compute.internal    Ready    <none>   8m9s    v1.15.10-eks-bac369


$ aws ec2 describe-instances --instance-ids i-0a1bbf530bdb47960 | grep SubnetId
                    "SubnetId": "subnet-0f008319d1124fe8a",
                            "SubnetId": "subnet-0f008319d1124fe8a",

$ aws ec2 describe-subnets --subnet-ids subnet-00a788781be10df8a
{
    "Subnets": [
        {
            "AvailabilityZone": "us-east-2a",
            "AvailabilityZoneId": "use2-az1",
            "AvailableIpAddressCount": 8176,
            "CidrBlock": "192.168.96.0/19",
            "DefaultForAz": false,
            "MapPublicIpOnLaunch": false,
            "State": "available",
            "SubnetId": "subnet-00a788781be10df8a",
            "VpcId": "vpc-060aa9359ca87b956",
            "OwnerId": "363193625107",
            "AssignIpv6AddressOnCreation": false,
            "Ipv6CidrBlockAssociationSet": [],
            "Tags": [
                {
                    "Key": "aws:cloudformation:stack-name",
                    "Value": "eksctl-attractive-creature-1587123638-cluster"
                },
                {
                    "Key": "alpha.eksctl.io/eksctl-version",
                    "Value": "0.18.0-dev+b8b1a818.2020-04-17T21:40:08Z"
                },
                {
                    "Key": "kubernetes.io/role/internal-elb",
                    "Value": "1"
                },
                {
                    "Key": "aws:cloudformation:logical-id",
                    "Value": "SubnetPrivateUSEAST2A"
                },
                {
                    "Key": "eksctl.cluster.k8s.io/v1alpha1/cluster-name",
                    "Value": "attractive-creature-1587123638"
                },
                {
                    "Key": "Name",
                    "Value": "eksctl-attractive-creature-1587123638-cluster/SubnetPrivateUSEAST2A"
                },
                {
                    "Key": "alpha.eksctl.io/cluster-name",
                    "Value": "attractive-creature-1587123638"
                },
                {
                    "Key": "aws:cloudformation:stack-id",
                    "Value": "arn:aws:cloudformation:us-east-2:363193625107:stack/eksctl-attractive-creature-1587123638-cluster/44ac80e0-80a0-11ea-8891-0a3e4885fce2"
                },
                {
                    "Key": "kubernetes.io/cluster/attractive-creature-1587123638",
                    "Value": "shared"
                }
            ],
            "SubnetArn": "arn:aws:ec2:us-east-2:363193625107:subnet/subnet-00a788781be10df8a"
        }
    ]
}

```

</details>

<details>
<summary>Create one new cluster with two managed node group (one public and one private</summary>

```shell script
$ cat test_private.yaml              
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: cluster-15
  region: us-east-2

managedNodeGroups:
  - name: managed-ng-public
    instanceType: m5.large
    minSize: 2
    desiredCapacity: 3
    maxSize: 4
    labels: {role: worker}
    tags:
      nodegroup-role: worker
    iam:
      withAddonPolicies:
        externalDNS: true
        certManager: true

  - name: managed-ng-private
    instanceType: m5.large
    # launch nodegroup in private subnets
    privateNetworking: true

$ ./eksctl create cluster -f test_private.yaml               
[ℹ]  eksctl version 0.18.0-dev+b8b1a818.2020-04-17T21:40:08Z
[ℹ]  using region us-east-2
[ℹ]  setting availability zones to [us-east-2c us-east-2a us-east-2b]
[ℹ]  subnets for us-east-2c - public:192.168.0.0/19 private:192.168.96.0/19
[ℹ]  subnets for us-east-2a - public:192.168.32.0/19 private:192.168.128.0/19
[ℹ]  subnets for us-east-2b - public:192.168.64.0/19 private:192.168.160.0/19
[ℹ]  using Kubernetes version 1.15
[ℹ]  creating EKS cluster "cluster-15" in "us-east-2" region with managed nodes
[ℹ]  2 nodegroups (managed-ng-private, managed-ng-public) were included (based on the include/exclude rules)
[ℹ]  will create a CloudFormation stack for cluster itself and 0 nodegroup stack(s)
[ℹ]  will create a CloudFormation stack for cluster itself and 2 managed nodegroup stack(s)
[ℹ]  if you encounter any issues, check CloudFormation console or try 'eksctl utils describe-stacks --region=us-east-2 --cluster=cluster-15'
[ℹ]  CloudWatch logging will not be enabled for cluster "cluster-15" in "us-east-2"
[ℹ]  you can enable it with 'eksctl utils update-cluster-logging --region=us-east-2 --cluster=cluster-15'
[ℹ]  Kubernetes API endpoint access will use default of {publicAccess=true, privateAccess=false} for cluster "cluster-15" in "us-east-2"
[ℹ]  2 sequential tasks: { create cluster control plane "cluster-15", 2 parallel sub-tasks: { create managed nodegroup "managed-ng-public", create managed nodegroup "managed-ng-private" } }
[ℹ]  building cluster stack "eksctl-cluster-15-cluster"
[ℹ]  deploying stack "eksctl-cluster-15-cluster"
[ℹ]  building managed nodegroup stack "eksctl-cluster-15-nodegroup-managed-ng-private"
[ℹ]  building managed nodegroup stack "eksctl-cluster-15-nodegroup-managed-ng-public"
[ℹ]  deploying stack "eksctl-cluster-15-nodegroup-managed-ng-public"
[ℹ]  deploying stack "eksctl-cluster-15-nodegroup-managed-ng-private"
[✔]  all EKS cluster resources for "cluster-15" have been created
[✔]  saved kubeconfig as "/home/tammach/.kube/config"
[ℹ]  nodegroup "managed-ng-public" has 3 node(s)
[ℹ]  node "ip-192-168-13-15.us-east-2.compute.internal" is ready
[ℹ]  node "ip-192-168-41-168.us-east-2.compute.internal" is ready
[ℹ]  node "ip-192-168-67-120.us-east-2.compute.internal" is ready
[ℹ]  waiting for at least 2 node(s) to become ready in "managed-ng-public"
[ℹ]  nodegroup "managed-ng-public" has 3 node(s)
[ℹ]  node "ip-192-168-13-15.us-east-2.compute.internal" is ready
[ℹ]  node "ip-192-168-41-168.us-east-2.compute.internal" is ready
[ℹ]  node "ip-192-168-67-120.us-east-2.compute.internal" is ready
[ℹ]  nodegroup "managed-ng-private" has 2 node(s)
[ℹ]  node "ip-192-168-105-50.us-east-2.compute.internal" is ready
[ℹ]  node "ip-192-168-170-75.us-east-2.compute.internal" is ready
[ℹ]  waiting for at least 2 node(s) to become ready in "managed-ng-private"
[ℹ]  nodegroup "managed-ng-private" has 2 node(s)
[ℹ]  node "ip-192-168-105-50.us-east-2.compute.internal" is ready
[ℹ]  node "ip-192-168-170-75.us-east-2.compute.internal" is ready
[ℹ]  kubectl command should work with "/home/tammach/.kube/config", try 'kubectl get nodes'
[✔]  EKS cluster "cluster-15" in "us-east-2" region is ready

$ kgno
NAME                                           STATUS   ROLES    AGE    VERSION
ip-192-168-105-50.us-east-2.compute.internal   Ready    <none>   96s    v1.15.10-eks-bac369
ip-192-168-13-15.us-east-2.compute.internal    Ready    <none>   105s   v1.15.10-eks-bac369
ip-192-168-170-75.us-east-2.compute.internal   Ready    <none>   88s    v1.15.10-eks-bac369
ip-192-168-41-168.us-east-2.compute.internal   Ready    <none>   90s    v1.15.10-eks-bac369
ip-192-168-67-120.us-east-2.compute.internal   Ready    <none>   103s   v1.15.10-eks-bac369

```

</details>

### Checklist
- [x] Added tests that cover your change (if possible)
- [x] Added/modified documentation as required (such as the `README.md`, and `examples` directory)
- [x] Manually tested
- [x] Added labels for change area (e.g. `area/nodegroup`) and target version (e.g. `version/0.12.0`)
- [ ] Added note in `docs/release_notes/draft.md` (or relevant release note)

### CODE REVIEW

**Comment by sayboras** [comment]
Have been using this to have the same feature as Console UI.

It also aligns with recent announcement as well https://aws.amazon.com/blogs/containers/upcoming-changes-to-ip-assignment-for-eks-managed-node-groups/.

So I am marking this PR ready for review, keen to hear your inputs though.

File: pkg/ctl/create/cluster_test.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+			Entry("with node-labels flag", "--node-labels", "partition=backend,nodeclass=hugememory"),
+			Entry("with node-zones flag", "--node-zones", "zone1,zone2,zone3"),
+			Entry("with asg-access flag", "--asg-access", "true"),
+			Entry("with external-dns-access flag", "--external-dns-access", "true"),
+			Entry("with full-ecr-access flag", "--full-ecr-access", "true"),
+			Entry("with appmesh-access flag", "--appmesh-access", "true"),
+			Entry("with alb-ingress-access flag", "--alb-ingress-access", "true"),
+		)
+
+		DescribeTable("with un-supported flags",
```
Comment (sayboras) [inline pkg/ctl/create/cluster_test.go:119]: This set of test cases are related to changes. Previously, _node-private-networking_ was not supported for managed node group.

**Comment by sayboras** [comment]
@cPu1 @martina-if For your input and review, happy to address any comments you have.

**Comment by martina-if** [comment]
:+1:  Thanks @sayboras LGTM. I would like @cPu1 to take a look as well

File: pkg/ctl/create/nodegroup_test.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+				Expect(err).To(HaveOccurred())
+				Expect(err.Error()).To(Equal(c.error.Error()))
+			},
+			Entry("without cluster name", invalidParamsCase{
+				args:  []string{"nodegroupName", "--name", "nodegroupName"},
+				error: fmt.Errorf("--cluster must be set"),
+			}),
+			Entry("with nodegroup name as argument and flag", invalidParamsCase{
+				args:  []string{"nodegroupName", "--cluster", "clusterName", "--name", "nodegroupName"},
+				error: fmt.Errorf("--cluster=nodegroupName and argument nodegroupName cannot be used at the same time"),
```
Comment (sayboras) [inline pkg/ctl/create/nodegroup_test.go:71]: This error mesage is wronged, existing GH issue https://github.com/weaveworks/eksctl/issues/1782. Will fix in subsequent PR.

File: pkg/ctl/create/nodegroup_test.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+			func(c invalidParamsCase) {
+				commandArgs := append([]string{"nodegroup", "--managed", "--cluster", "clusterName"}, c.args...)
+				cmd := newDefaultCmd(commandArgs...)
+				_, err := cmd.execute()
+				Expect(err).To(HaveOccurred())
+				Expect(err.Error()).To(Equal(c.error.Error()))
+			},
+			Entry("with nodegroup name as argument and flag", invalidParamsCase{
+				args:  []string{"nodegroupName", "--name", "nodegroupName"},
+				error: fmt.Errorf("--cluster=nodegroupName and argument nodegroupName cannot be used at the same time"),
```
Comment (sayboras) [inline pkg/ctl/create/nodegroup_test.go:143]: same as above

**Comment by cPu1** [comment]
I ran the integration tests for managed nodegroups and it was successful. Thanks for the PR!

**Comment by cPu1** [comment]
I've tweaked the title to make this feature clear in the release notes.

**Comment by sayboras** [comment]
@martina-if @cPu1 thanks for your review, I made some changes for integration test as well, will verify one more time tomorrow as well. :100:

**Comment by sayboras** [comment]
Today integration test failed for 2 cases, but they are related to gitops I think. Please let me know if you have other idea.


https://app.circleci.com/pipelines/github/weaveworks/eksctl/1419/workflows/a7582011-c09a-4d3d-9454-aa85e28acbea/jobs/7442

```
...
[6] Summarizing 2 Failures:
[6] 
[6] [Fail] Enable and use GitOps quickstart profiles enable repo [It] should add Flux to the repo and the cluster 
[6] /src/integration/tests/quickstart_profiles/quickstart_profiles_test.go:72
[6] 
[6] [Fail] Enable and use GitOps quickstart profiles enable profile [It] should add the configured quickstart profile to the repo and the cluster 
[6] /src/integration/matchers/flux.go:48
[6] 
[6] Ran 2 of 2 Specs in 1123.094 seconds
[6] FAIL! -- 0 Passed | 2 Failed | 0 Pending | 0 Skipped
[6] --- FAIL: TestSuite (1123.10s)
[6] FAIL
...

...
2020/04/21 04:32:19 ================== Summary ================================
[1] Ran 4 of 4 Specs in 720.909 seconds
[1] SUCCESS! -- 4 Passed | 0 Failed | 0 Pending | 0 Skipped
[1] --- PASS: TestSuite (720.91s)
[1] PASS
[1] 
[1] Ginkgo ran 1 suite in 12m47.597391827s
[1] Test Suite Passed
[7] Ran 1 of 1 Specs in 988.523 seconds
[7] SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
[7] --- PASS: TestSuite (988.52s)
[7] PASS
[7] 
[7] Ginkgo ran 1 suite in 17m11.985628662s
[7] Test Suite Passed
[6] Ran 2 of 2 Specs in 1123.094 seconds
[6] FAIL! -- 0 Passed | 2 Failed | 0 Pending | 0 Skipped
[6] --- FAIL: TestSuite (1123.10s)
[6] FAIL
[6] 
[6] Ginkgo ran 1 suite in 19m28.732130159s
[6] Test Suite Failed
[0] Ran 1 of 1 Specs in 1508.055 seconds
[0] SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
[0] --- PASS: TestSuite (1508.06s)
[0] PASS
[0] 
[0] Ginkgo ran 1 suite in 25m51.906750671s
[0] Test Suite Passed
[5] Ran 13 of 13 Specs in 1761.134 seconds
[5] SUCCESS! -- 13 Passed | 0 Failed | 0 Pending | 0 Skipped
[5] --- PASS: TestSuite (1761.14s)
[5] PASS
[5] 
[5] Ginkgo ran 1 suite in 30m8.785996538s
[5] Test Suite Passed
[3] Ran 35 of 35 Specs in 2261.072 seconds
[3] SUCCESS! -- 35 Passed | 0 Failed | 0 Pending | 0 Skipped
[3] --- PASS: TestSuite (2261.07s)
[3] PASS
[3] 
[3] Ginkgo ran 1 suite in 38m31.462258708s
[3] Test Suite Passed
[2] Ran 10 of 10 Specs in 3659.053 seconds
[2] SUCCESS! -- 10 Passed | 0 Failed | 0 Pending | 0 Skipped
[2] --- PASS: TestSuite (3659.05s)
[2] PASS
[2] 
[2] Ginkgo ran 1 suite in 1h1m46.578081084s
[2] Test Suite Passed
[4] Ran 3 of 3 Specs in 5013.825 seconds
[4] SUCCESS! -- 3 Passed | 0 Failed | 0 Pending | 0 Skipped
[4] --- PASS: TestSuite (5013.83s)
[4] PASS
[4] 
[4] Ginkgo ran 1 suite in 1h24m18.076345719s
[4] Test Suite Passed
```

**Comment by xuhualin99** [comment]
Does this feature still require MapPublicIpOnLaunch set to true on the public subnet? I'm trying to create a managed ng in private subnet and my cluster has both public and private subnet. Thanks

**Comment by xuhualin99** [comment]
Never mind, it turns out I also need to enable private access for API end point.

**Pull request merged**